docs/fsharp/language-reference/type-extensions.md

```diff
@@ -1,99 +1,151 @@
+An optional type extension is an extension that appears outside the original module, namespace, or assembly of the type being extended.
 
-You can use intrinsic type extensions to separate the definition of a type into sections. This can be useful in managing large type definitions, for example, to keep compiler-generated code and authored code separate or to group together code created by different people or associated with different functionality.
+Optional type extensions are useful for extending a type that you have not defined yourself. For example:
 
-In the following example, an optional type extension extends the `System.Int32` type with an extension method `FromString` that calls the static member `Parse`. The `testFromString` method demonstrates that the new member is called just like any instance member.
+```fsharp
+module Extensions
 
-[!code-fsharp[Main](../../../samples/snippets/fsharp/lang-ref-2/snippet3702.fs)]
+open System.Collections.Generic
 
-The new instance member will appear like any other method of the `Int32` type in IntelliSense, but only when the module that contains the extension is open or otherwise in scope.
+type IEnumerable<'T> with
+    /// Repeat each element of the sequence n times
+    member xs.RepeatElements(n: int) =
+        seq {
+            for x in xs do
+                for i in 1 .. n do
+                    yield x
+        }
+```
 
-## Generic Extension Methods
-Before F# 3.1, the F# compiler didn't support the use of C#-style extension methods with a generic type variable, array type, tuple type, or an F# function type as the "this" parameter. F# 3.1 supports the use of these extension members.
+You can now access `RepeatElements` as if it were a member on <xref:System.Collections.Generic.IEnumerable`1> so long as the `Extensions` module is opened in the scope that you are working in.
 
-For example, in F# 3.1 code, you can use extension methods with signatures that resemble the following syntax in C#:
+Optional extensions do not appear on the extended type when examined by reflection. Optional extensions must be in modules, and they are only in scope when the module that contains the extension is open or is otherwise in scope.
 
-```csharp
-static member Method<T>(this T input, T other)
-```
+Optional extension members are compiled to static members for which the object instance is passed implicitly as the first parameter. However, they act as if they were instance members or static members according to how they are declared.
 
-This approach is particularly useful when the generic type parameter is constrained. Further, you can now declare extension members like this in F# code and define an additional, semantically rich set of extension methods. In F#, you usually define extension members as the following example shows:
+It is not possible to define an optional extension on a generic type when the type variable is constrained. For example, the following is not possible:
```

Comment:
So this is actually possible, and I use it a lot. only apparently has some restrictions for the static member extension declarations

Reply:
Seems like it's not just for static member extension declarations:

```fsharp
type C<'T when 'T : null>() =
    member __.Data =
        Array.zeroCreate<'T> 12
        |> Seq.ofArray

type C<'T> with
    member this.Bananas() = Seq.sum this.Data
```

This gives:

```
This code is not sufficiently generic. The type variable  ^T when  ^T : null and  ^T : (static member get_Zero : ->  ^T) and  ^T : (static member ( + ) :  ^T *  ^T ->  ^T) could not be generalized because it would escape its scope.
```

Which is sort of what I'd expect.

Also, if you do not match the type constraints of the declaration and type extension, then you get a warning:

```
One or more of the declared type parameters for this type extension have a missing or wrong type constraint not matching the original type constraints on 'C<_>'
```

So the rules seem to be making sure that you match constraints with the declared type _and_ that you don't write code that ... confuses the type checker?